### ISSUE TITLE
[functional-parameters] ignore pattern option not working with object expressions

### ISSUE THREAD

**Issue by MrZyr0** [jonaskello/eslint-plugin-functional#282]
# Bug Report

I am trying to use the [functional-parameters](https://github.com/jonaskello/eslint-plugin-functional/blob/master/docs/rules/functional-parameters.md) rule in a Nuxt.Js 2 project.
As you may know, Vue.Js (on which Nuxt.Js is based) provides functions for its lifecycle like `mounted()`.

## Expected behavior

These functions will have no parameters in most cases, so I have to ignore them.

## Actual behavior

Unfortunately, despite many attempts, I was unable to ignore them.

## Steps to reproduce

Here is the last configuration I tried for the rule:

```json
{
  "rules": {
    "functional/functional-parameters": [
			"error",
			{
				"ignorePattern": [
					"^mounted$",
					"^created$",
					"^unmount$",
					"^unmounted$",
					"^beforeDestroy$",
					"^destroy$",
					"^updated$",
					"^beforeUpdate$",
					"^onBeforeUpdate$",
					"^onUpdated$",
					"^onMounted$"
				]
			}
		]
  }
}
```

## Proposed changes

- [ ] Fix `ignorePattern` property for this rule
- [ ] Propose some ignorance models according to frameworks in the documentation

**Comment by MrZyr0** [comment]
## You are awesome !

I've updated the plugin, it seems to resolve the error of vue lifecycle functions.

The last thing to fix for the `ignorePattern` is for some functions like `console.log` for the development environment.

I also try for `setTimeout` and I don't work :

<img width="1000" alt="image" src="https://user-images.githubusercontent.com/32927775/140647531-8e7f9a1e-b9dd-4951-989d-5701bb9f821e.png">


Here's my `.eslintrc` :
```json
{
	"root": true,
	"env": {
		"browser": true,
		"node": true,
		"es2021": true
	},
	"parserOptions": {
		"ecmaVersion": 2021,
		"ecmaFeatures": {
			"generators": false,
			"objectLiteralDuplicateProperties": false
		},
		"parser": "@babel/eslint-parser",
		"requireConfigFile": false
	},
	"extends": [
		"airbnb/base",
		"plugin:jsonc/base",
		"plugin:unicorn/recommended",
		"@nuxtjs",
		"plugin:nuxt/recommended",
		"plugin:functional/external-recommended",
		"plugin:functional/all",
		"plugin:sonarjs/recommended",
		"plugin:pii/recommended",
		"prettier"
	],
	"plugins": ["write-good-comments", "functional", "no-secrets", "sonarjs"],
	"rules": {
		"functional/no-expression-statement": ["error", { "ignoreVoid": true }],
		"functional/functional-parameters": [
			"error",
			{
				"ignorePattern": [
					"^mounted.*",
					"^created.*",
					"^unmount.*",
					"^unmounted.*",
					"^beforeDestroy.*",
					"^destroy.*",
					"^updated.*",
					"^beforeUpdate.*",
					"^onBeforeUpdate.*",
					"^onUpdated.*",
					"^onMounted.*",
					"^console.info.*",
					"^setTimeout.*"
				]
			}
		]
	}
}
```

**Comment by RebeccaStevens** [comment]
That's a different rule that needs to be configured:

```json
"functional/no-expression-statement": [
    "error",
    {
        "ignoreVoid": true,
        "ignorePattern": [
            "^setTimeout"
        ],
    },
],
```

I'm not sure why `console.info` is reporting a violation. That suggests that type information isn't working as the `ignoreVoid` option should handle it. You can ignore it manually by adding this pattern `"^console\\.info"`.

For the list of functions you're passing to `ignorePattern` for `functional-parameters`, there is nothing that comes after the function's name so you don't need to add `".*"` to the end unless you want include all functions that are prefixed with that string.

Also, remember to escape `"."` when using it as a character in a regex string.

**Comment by MrZyr0** [comment]
Thank you so much for your awesome work !

Indeed, I made a mistake in the rules for the configuration 😅 

Now it's working 🤩 

However, as you said, it seems that `ingoreVoid` does not work. 😢

**Comment by RebeccaStevens** [comment]
`ignoreVoid` will work as long as type information is available to eslint. make sure your config includes:

```js
parserOptions: {
    tsconfigRootDir: __dirname, // this line is optional and only works if using JS config file.
    project: "tsconfig.json",
},
```

**Comment by MrZyr0** [comment]
Humm 🤔

Okay, here is my `parserOptions` :
```js
"parserOptions": {
    "ecmaVersion": 2021,
    "ecmaFeatures": {
        "generators": false,
        "objectLiteralDuplicateProperties": false
    },
    "parser": "@babel/eslint-parser",
    "requireConfigFile": false
},
```

Since, I don't use typescript I have only a `jsconfig.json` :
```json
{
  "compilerOptions": {
    "target": "esnext",
    "module": "es6",
    "checkJs": true,
    "esModuleInterop": true, // force import
    "allowSyntheticDefaultImports": true, // force import
		"experimentalDecorators": true,
    "baseUrl": ".",
    "paths": {
      "~/*": ["./*"],
      "@/*": ["./*"],
      "~~/*": ["./*"],
      "@@/*": ["./*"]
    },
  },
	"checkJs": true,
  "exclude": [".husky", ".nuxt", "node_modules", "dist"]
}
```

Est-ce-que je peux quand même faire quelque chose ?

**Comment by RebeccaStevens** [comment]
Ah, if you're not using typescript then you can't use the `ignoreVoid` option. That option requires type information in order to identify what functions are void returning functions. You'll need to manually specify the names of all the functions when just using JavaScript.

**Comment by MrZyr0** [comment]
I've detected one problem with `functional/functional-parameters` in vue.
We use simple arrow function many times to set default value of props or computed.


As this arrow function is only used to set a value, 99% we don't need a parameter. This triggers the rule and creates a false positive. 😕

<img width="992" alt="image" src="https://user-images.githubusercontent.com/32927775/141645318-44589d39-3470-4f2e-b8e2-9a6032f3c1c7.png">

**Comment by MrZyr0** [comment]
> Ah, if you're not using typescript then you can't use the `ignoreVoid` option. That option requires type information in order to identify what functions are void returning functions. You'll need to manually specify the names of all the functions when just using JavaScript.

Ok, no problem. I only need to know. :) 

I think it can be useful to specify this in documentation.

**Comment by RebeccaStevens** [comment]
https://github.com/jonaskello/eslint-plugin-functional/blob/master/docs/rules/no-expression-statement.md#ignorevoid

**Comment by RebeccaStevens** [comment]
> As this arrow function is only used to set a value, 99% we don't need a parameter. This triggers the rule and creates a false positive. 😕

It's not a false positive as that's exactly what the rule is designed to do when `enforceParameterCount` isn't set to `false` (`false` is not the default).
https://github.com/jonaskello/eslint-plugin-functional/blob/master/docs/rules/functional-parameters.md#enforceparametercount

**Comment by MrZyr0** [comment]
Hmmm ok you are right, but it is a problem with this rule anyway 😕
